### DISCUSSION TITLE
Getting up and running

### DISCUSSION THREAD

**Discussion by Sternbach-Software** [Sternbach-Software/Wabbitemu-Android#2] [General]
Continuing from old repo

**Comment by Sternbach-Software** [comment]
I wanted to update the tooling, which required bumping the minimum supported Android version from API 16 (Android 4.1) to API 21 (Android 5.0). Is that OK?

**Reply by cheintz** [reply]
Fine by me, I think anyone still running Android <5.0 can just use the old one, since it won't be running on the 64 bit only OS/hardware.

**Reply by thecomet25** [reply]
Yea I can't see that being an issue

**Reply by DanHolli** [reply]
> Yea I can't see that being an issue

Even if it were an issue, there's still archives of Wabbit spread across the internet that still run fine on those older versions

**Comment by ddm-j** [comment]
The hero returns. The prophecy will be fulfilled! I will also tithe $25 to see WabbbitEmu live again.

**Reply by Sternbach-Software** [reply]
Thank you so much! You can support me through my patreon! https://www.patreon.com/posts/wabbitemu-147583158

**Reply by Sternbach-Software** [reply]
@ddm-j

**Comment by johanw666** [comment]
Nice. I have some old ROM's if anyone is interested:

TI-83+ v1.13 (2001)(Texas Instruments)[Update].8xu
ti-84ce.rom
TI-89 Hardware 1 v2.03 (1999)(Texas Instruments).rom
TI-89 Hardware 2 v2.03 (1999)(Texas Instruments).rom
TI-89 Hardware 2 v2.05 (2000)(Texas Instruments)[Update].89u
TI-89 Hardware 2 v2.09 (2000)(Texas Instruments).89u
TI-89_Titanium_OS-3.10.89u
TI-92+ Hardware 2 v2.04 (2000)(Texas Instruments).rom
ti73-1.19.rom
TI82.rom
TI84PlusSE.rom
ti84se.rom
ti85.zip
TI86.rom
ti89.pdf
ti89.rom
ti92p.rom
TI92plus-os-2.09.9xu
Voyage200-os-3.10.v2u

**Comment by Sternbach-Software** [comment]
Making good progress! I think I got past the startup screens, so the user selects the model and ROM and then it reads the ROM file correctly. 

I have to double check, but it may only work if you give the app permission to access all files 🫤 because of Android permissions and the limitations of reading files from C code on Android. It copies the file to the app sandbox/internal files before it goes to C, and then reads the copied file from C.

It goes to the calculator screen, and lays out the buttons in the right order, but I have to create a mapping from the key code to what it says on the calculator, and add the super/sub text for the shift functions (right now it's a grid of buttons with some cells empty). I also have to figure out why the LCD screen is totally black and doesn't seem to respond to the buttons 🫤 not sure if I should try to revert a bunch of code back to the un-modernized version before I changed anything and add in the parts that were necessary, or try to debug and fix the problem.

I also need to create the layouts for the rest of the models, right now I'm working with the ti-84 SE (my fav) as a test.

**Reply by thecomet25** [reply]
On the standard TI, sometimes the screen goes black and a hard reset or quit will clear it. I just know I saw that same thing on my physical and the old app. So it may not be the code. 

I'm a fan of the 84CSE myself, same as the 84, but _color_ lol

**Comment by johanw666** [comment]
but it may only work if you give the app permission to access all files 🫤 because of Android permissions and the limitations of reading files from C code on Android.

I have no objection against that, I use that permission in a project of myself too to make things easier, but Google is strict with that permission when adding files to the Play Store (it says it basically only permits that for file managers etc.) so if that is your goal you might have to rethink that. You can ask permission to access one selectable directory since, afaik, Android 8 (below you can use the READ_EXTERNAL_STORAGE permission).

**Reply by Sternbach-Software** [reply]
Yeah, I know. I figured it would just be an APK? Would be a shame. But I couldn't figure out how to use asset manager or content resolvet from the NDK...

**Reply by thecomet25** [reply]
The original was pulled from the store a while back. It may have to stay off the play store regardless, especially as we're simply adapting the old source. Not sure how that works legally

**Reply by ForrestHerman** [reply]
> The original was pulled from the store a while back. It may have to stay off the play store regardless, especially as we're simply adapting the old source. Not sure how that works legally

Excuse my ignorance, but is there a reason it couldn't go on F-Droid?

**Reply by Sternbach-Software** [reply]
I don't see why it can't go on f-droid. Didn't think about the rights issue.

**Reply by DanHolli** [reply]
> The original was pulled from the store a while back. 

I'm pretty sure it was pulled because it wasn't updated for newer versions of android. But I dont think it would be an issue with a new package name (to not conflict with the old one)

**Comment by cheintz** [comment]
It could maybe go on Izzy ondroid.

More and more people (including me) are also using Obtainium to manage and
update APKs that are distributed through github releases.

**Comment by cheintz** [comment]
Any news? I wish I could help, but have no android app development experience...

**Comment by Sternbach-Software** [comment]
@cheintz 
# Update 
I've been thinking all week about how I want to deal with the calculator screen going forward. On the one hand, it's already partially working using the old technology (Android View). However, the buttons would be easy to do in the new technology (Jetpack Compose). I'm kind of undecided. I will probably give in to inertia and use the old technology.

@cheintz if you want to help, create a  structured list of every button on every calculator model besides 84SE. If there are 6 buttons in a row, and only a button at the end, the row should be `null,null,null,null,null,KeyEvent.KEYCODE_DPAD_UP`. It is basically mapping the [C code layouts](https://github.com/Sternbach-Software/Wabbitemu-Android/blob/master/app%2Fsrc%2Fmain%2Fjni%2Fhardware%2Fkeys.c) to [KeyEvent](https://developer.android.com/reference/android/view/KeyEvent) keycode constants. Here is the [84SE example](https://github.com/Sternbach-Software/Wabbitemu-Android/blob/master/app/src/main/java/io/github/angelsl/wabbitemu/calc/variants/TI84PSELayout.kt). Also some kind of structured mapping of key to shift/alpha/secondary function (even just the label text, though the KeyEvent would be nice as well) would be helpful.


I still need to figure out why the screen doesn't turn on. But otherwise, I assume it's ready for public consumption.

**Reply by thecomet25** [reply]
TI-83 and TI-84 each have the same button layouts correct? Just shifted up a bit on the edges. I'll be honest, outside of the TI-89 Titanium I wasn't even aware of the others from the ROM list above. Had no idea those were ever supported.

**Comment by Sternbach-Software** [comment]
I am starting to think that implementing calculators like this (including button labels) is not a good idea:
![Screenshot 2024-07-06 at 9 02 29 PM](https://github.com/Sternbach-Software/Wabbitemu-Android/assets/60489781/c1507d0e-2d91-45ac-bf73-1775dd2e1a60)
Partially because of the contextual memory of the calculator skin, and partially because I still have to reverse engineer how the original author meant for these images to be used in conjunction with key codes (I thought he mapped how much R, G, and B was in each pixel to a hit box and key code, but I apparently didn't do it right):
![ti84psekeymap](https://github.com/Sternbach-Software/Wabbitemu-Android/assets/60489781/2d804ca7-9b2c-4b08-a682-2c53af4ebc26) 
![ti84pse](https://github.com/Sternbach-Software/Wabbitemu-Android/assets/60489781/9beb613c-bfa1-442c-802a-4c0f8cd7c8a6)

**Reply by thecomet25** [reply]
Any luck figuring out the button mapping? Anything we can do to assist or brainstorm with you on?

**Comment by TrashySat** [comment]
Many Thanks for the development 🎉🎊🎁, my TI-89 missing me so much on Android since last Month !!!
I follow this post day after day 😊

**Comment by thecomet25** [comment]
@Sternbach-Software how's it looking? Any version available to try or anything? Hate to see this project fizzle out. Still happy to follow through with the bounty if we can get something up and running

**Reply by Sternbach-Software** [reply]
I think I've gotten decision fatigue on the questions I posed to the "community". I still think about the project every Sunday, but all of the unanswered questions are overwhelming - especially when it's unclear what people will want, and there are too many paths forward.

**Reply by thecomet25** [reply]
May be my newness to Github, but I feel like the general gist as far as I can tell is to pick a standard ROM, let's say the 84SE as that's the most common calc on the market and just go with it. I worry that trying to allow and adjust every single calc design is too much. 

Speaking for myself, having the skin that maintains the look of the 84 is important, but how those buttons are implemented is up to you. 


What are the other open questions right now. Maybe aggregating them in one place to find answers for could be helpful.

**Comment by Sternbach-Software** [comment]
I wonder if the code ever needed changing in the first place. The code to download the roms didn't work for sure, but maybe the code just needed to be recompiled using the latest versions of the Android build tools and the target Android version increased...

I'm going to try running the old version, and maybe replace the startup wizard with the new code, and see what happens. Might have to add Kotlin language support because the new wizard was written in Kotlin, but I might have to translate it to legacy Java.

**Comment by Sternbach-Software** [comment]
Or maybe the reason the calculator part isn't working now is because of the code incompatibility with a newer version of Android? I've been testing using Android API 34 (Android 14)

**Comment by Sternbach-Software** [comment]
I got the old emulator code working (with the exception of button clicks sometimes being ignored... 🤔) on Android 14 with the new ROM selection code and build tools, but the scaling is off. The highlighted button is the parenthesis button:
![Screenshot 2024-08-10 at 2 54 39 AM](https://github.com/user-attachments/assets/b51dddb8-ea22-43a3-b313-7736ccc83d7a)

**Comment by Sternbach-Software** [comment]
I'm wondering if I scale the blueprint images for different screen sizes whether the scaling issue will be fixed. https://github.com/Sternbach-Software/Wabbitemu-Android/discussions/2#discussioncomment-9977068

**Comment by Sternbach-Software** [comment]
Can someone help with the scaling of the images? They are contained in [this](https://github.com/Sternbach-Software/Wabbitemu-Android/tree/master/app/src/main/assets) folder. The ratios for scaling them up/down for different sized devices are [here](https://developer.android.com/training/multiscreen/screendensities#TaskProvideAltBmp). @thecomet25 maybe?

**Reply by thecomet25** [reply]
I'll take a look. Happy to help. 

That said, I remember this being an issue on my phone in a past version. I believe there was a checkbox or setting on the menu where you select a ROM. I have an old phone somewhere, I'll try loading it there and seeing if I can recreate it. It hopefully is just that simple

**Reply by thecomet25** [reply]
Yep! So just checked, in the settings menu from the menu you can pull in from a swipe from the left, there is a setting called Fix Aspect Ratio as well as a setting for smaller devices and larger devices. Try that, but if it's still an issue I can absolutely help with image scaling. I'm much better at image manipulation than coding!

**Reply by thecomet25** [reply]
Also, can you include your current apk in the repository? Tried compiling it on my own and definitely messed something up lol. 

Pretty confident the lefthand menu has the scaling solution you need though. Hopefully good news!

**Reply by Sternbach-Software** [reply]
Seems like its the same issue on screens of varying pixel density and ratios (though I obviously didn't try all possible configurations)

**Reply by Sternbach-Software** [reply]
Tried fixing the aspect ratio, didn't work

**Comment by johanw666** [comment]
Whatever the discussions about the gui side, if the goal is to get it back into Google Play the native libraries will have to be available in 64 bit too. They are now only 32 bit.

**Reply by Sternbach-Software** [reply]
I'm not really a C/C++ guy, so I don't fully understand the implications of this. I guess I will just try compiling to a 64-bit device and see if it works? There's no translation layer? There's got to be. Or maybe the code needs to be re-written? Not 100% sure where to start on this one.

**Comment by cheintz** [comment]
Thank you for your continued efforts!

My main ask is to be able to use Wabbitemu with TI83/84 on Pixel 7 pro, ideally under GrapheneOS, which is 64 bit only.

**Reply by Sternbach-Software** [reply]
Got it. Not sure I will easily be able to test GrapheneOS in a development environment though...

**Comment by Sternbach-Software** [comment]
@thecomet25 All, [here](https://github.com/Sternbach-Software/Wabbitemu-Android/raw/master/app-release.apk) is a signed APK of the code I am working with.

The ROM selection is the new code, the calculator is mostly the old code (I think there are some internal updates there too like newer library versions etc., maybe some code cleanups or something too, would have to check). However, the scaling is off for me, as I mentioned here. So, if button hit-boxes are in the right place for you (clicking on the image for "5" enters 5 on the LCD), _please_ report what device you used or what settings you changed to make it work. I only have a few more options I can think of to try to fix the old code with minimal changes, and if they don't work, I think the old code will have to be scrapped.

**Reply by thecomet25** [reply]
Hmm.. Yep, still seeing the same issues you are. 

Few things I've noticed:
1):  The buttons aren't fully responsive, I'll hit them and they are off-center, but they also only respond once or so to being hit, I can't repeatedly hit a button.
2): The header at the top stays on. At the top of my screen , and yours judging by the screenshots, it says Wabbitemu in a large title across the screen. This isn't the same as how it used to be, it used to be fully immersive and fullscreen. I can't be 100% but I think this is about the offset that we are seeing the button presses. 
3) Color picker seems to work great 
4) The back button instead of exiting the app takes you back to the ROM picker. Not the worst thing, just odd. The Load File and Re-run Setup Wizard buttons also seem to use the old ROM picker and are not functional

Not trying to overwhelm with things that don't work or do. Obviously just getting it running is important. But hoping that listing everything may help find a solution

**Reply by Sternbach-Software** [reply]
@thecomet25 oh, number 2 is really interesting and a potential clue. Will look into it. Definitely helpful.

**Reply by Sternbach-Software** [reply]
Yeah, I haven't touched the other parts you mentioned.

**Reply by Sternbach-Software** [reply]
@cheintz can you download and run the APK?

**Reply by Sternbach-Software** [reply]
@johanw666 can you download and run the APK on your 64 bit device?

**Reply by thecomet25** [reply]
Tried the app on my emulator and had the same issue, so repeatable on BlueStacks on Windows and 64-bit Galaxy S24

**Reply by johanw666** [reply]
> @johanw666 can you download and run the APK on your 64 bit device?

It runs, although it doesn't recognize all my rom files. For some roms it just returns to thr rom selector. My TI84 rom works.

Further, a menu tries to come in from the left when I use the menu button (unavailable on a kodern Android but available via the One Hand Operation app from Samsung that is originally an availability app). However, the menu shows no items but when I tap on it it tries to do something.

**Reply by Sternbach-Software** [reply]
@johanw666 does it display any error snackbar/toast message when you select a "bad" ROM?

**Reply by Sternbach-Software** [reply]
@thecomet25 @johanw666 would it be feasible to distribute the APK instead of going through an app store/distributer? How easy was it to install the app? Did you have to change any security settings on your phone to allow APK installation? Was it intuitive?

**Reply by johanw666** [reply]
> @johanw666 does it display any error snackbar/toast message when you select a "bad" ROM?

I get a toast "Please select a .rom or .sav file".

**Reply by johanw666** [reply]
> @thecomet25 @johanw666 would it be feasible to distribute the APK instead of going through an app store/distributer? How easy was it to install the app? Did you have to change any security settings on your phone to allow APK installation? Was it intuitive?

I'm an atypical user, installing apk's is the normal way for me to operate. I have disabled the Play Store and download (no install) apk's with a modified Aurora Store.

**Reply by thecomet25** [reply]
For me, no security settings needing changing other than "allow install from Chrome". You just need to have a ROM file that is ripped from a genuine TI84. That's the only difficult thing because it's the file that is protected. 

That said if one were to need to look up "TI84 ROM" somewhere, the 3rd result by the Internet Archive would likely be very helpful

**Comment by Sternbach-Software** [comment]
Good news everyone! Thanks to @thecomet25 ([here](https://github.com/Sternbach-Software/Wabbitemu-Android/discussions/2#discussioncomment-10312173)), we figured out the scaling issue! The toolbar was offsetting the buttons. There's still some quirky behavior occasionally with button clicks (e.g. pressing delete in the below screenshot makes it glitch and redraw the same screen on the LCD a bunch of times, ignored clicks, and maybe one or two others), but I think they might be solved by reverting some code to a previous version. Does anyone know what version of the app was working perfectly on their device?

Notice that the parenthesis button is properly aligned because of the missing toolbar:
![Screenshot 2024-08-17 at 12 02 15 PM](https://github.com/user-attachments/assets/f181976a-f195-4f82-af3a-67e2ddc570d2)

**Reply by thecomet25** [reply]
https://github.com/BuckeyeDude/Wabbitemu

This apk was working for me, about 95% it was that one. I believe it is the main Android port.

**Comment by Sternbach-Software** [comment]
Figured out why keys are being ignored. Need to figure out how to fix it.

**Reply by thecomet25** [reply]
Was sounding there for a while like you were making some solid headway. How are things going? Anything we can do to help Sternbach?

**Reply by cheintz** [reply]
I still had the offset and keys ignored issue using the apk in this repo (1.06.0) in Grapheneos with all the app compatibility stuff as lenient as possible. I have a screen capture with clicks and an app log if that would be helpful.

**Comment by thecomet25** [comment]
Project seems dead again. My bounty still stands.

**Reply by cheintz** [reply]
As does mine. It feels like we might be close...

**Reply by Sternbach-Software** [reply]
@cheintz we are close. The bug seems relatively simple: when a key is pressed, there is a function that keeps calling itself until it registers that the key was pressed and the action executed. It runs in an infinite loop because at some place in the code, the action is (sometimes) never registered properly. It's unclear where it isn't being "registered". It's a lot of code, and not documented at all. My next attempt was going to be checking out the fork that someone said was working, and just replacing all the C code of this fork with that one, and seeing if it is fixed.

The only changes I made are changing some stuff so the app builds using modern build tools, and updated the file picker (and reader) code because of the updated Android file permissions structure.

**Reply by thecomet25** [reply]
Thanks for the update @Sternbach-Software! Hoping for the best. Undocumented code is awful to parse through for sure

**Reply by cheintz** [reply]
Thanks for the update and for your hard work! @Sternbach-Software!

**Reply by cheintz** [reply]
Hi @Sternbach-Software , just thought I'd check in. My bounty stands.

**Reply by thecomet25** [reply]
Double bump. And agree

**Reply by chaos1267** [reply]
Created an account just to bump. Hope you come back to this. I'd seriously pay 100 dollars for a copy if you got this app updated and working again.

**Reply by DanHolli** [reply]
Sternbach probably is busy with life, these comments aren't very helpful.
If you want to be helpful, please familiarize yourself with the code, test and report issues and/or create pull requests.
If you truly believe it is dead you can create a fork and find people to work on it

**Reply by Sternbach-Software** [reply]
@DanHolli I very much see these comments, and it is very much alive in my mind. I've been in a several month sprint at work to keep my job (in this economy?). Looks like I'm about to get out of a 9-month tunnel, so hopefully I'll finish up the remaining tasks here. I think about this project almost every weekend.

**Reply by Sternbach-Software** [reply]
AI may have figured out the issue. Gonna test now.

**Reply by Sternbach-Software** [reply]
Might use AI to update the whole codebase

**Reply by Sternbach-Software** [reply]
@chaos1267 @thecomet25 @cheintz You can support me on my patreon! The app is out in the play store! 64bit compliant, 16Kb page size compliant, newest version of android, etc.! https://www.patreon.com/posts/wabbitemu-147583158

**Comment by Sternbach-Software** [comment]
@cheintz @DanHolli @chaos1267 @thecomet25 @johanw666 @TrashySat @ForrestHerman @ddm-j We are back in business! AI (Google Jules with Gemini 3) fixed the infinite looping issue! Now I need to make it 16KB page file size compatible because of the new Google Play policy, and possibly fix some slight UI misalignment (not perfectly aligned, but now it works).

**Reply by Sternbach-Software** [reply]
Immersive mode moves the hit boxes below the buttons on my OnePlus 8T, but if you turn it off, they're perfectly aligned.

**Reply by Sternbach-Software** [reply]
Email software@sternbach.org to join the Android internal testing track

**Reply by Sternbach-Software** [reply]
In review for open testing

**Reply by thecomet25** [reply]
@Sternbach-Software I couldn't have asked for a better Christmas present. Hopefully the testing is successful and happy to be a part of it. Once we verify it works for our devices make sure to get us your Venmo or equivalent so we can show you the love you deserve!

**Reply by cheintz** [reply]
This is excellent news! Thank you for your efforts!

**Comment by Sternbach-Software** [comment]
The app is public!! https://play.google.com/store/apps/details?id=org.sternbach.software.wabbitemu&pli=1

I guess with every release I can include an APK in this repository in case the app is ever taken down.

**Reply by thecomet25** [reply]
LEGEND! Just tried it and it works incredibly well. Thank you so much. What's your preferred form of payment to claim the bounty?

**Reply by mikepierce** [reply]
This is great! :) An APK release here would be appreciated

**Reply by Sternbach-Software** [reply]
@mikepierce https://github.com/Sternbach-Software/Wabbitemu-Android/releases/tag/1.0.0%2B6

**Reply by cheintz** [reply]
Excellent! It works on GrapheneOS on Pixel 7 pro. I had to enable "exploit protection compatibility mode" to select the rom on one profile, but that's the only issue I had.
